### PR TITLE
US #646: Task #819: bulk_update_kanban_order support swimlanes

### PR DESCRIPTION
**Pull request by bameda**
This PR change `/userstories/bulk_update_kanban_order` to simplify and support swimlanes. Now you have to call this endpoint with this data:

```
{
    "project_id": <project_id>,
    "status_id": <status_id>,
    "swimlane_id":<swimlane_id | null>,
    "after_userstory_id": <userstory_id | null>,
    "bulk_userstories": [<us1_id>, <us2_id>...]
 }
```

NOTE: if `after_userstory_id` is set, fields status_id, project_id and swimlane_id must be the same of this user story to prevent data errors.

 
### How To Review:

 - [x] Prerequisites
   - [x] Regenerate data with sample data.
   - [x] Serve branch task/870/new-dd-api from `taiga-front`
- [x] Test 0: Review
  - [x] Review the code
  - [x] Review the test code to understand how it works
  - [x] Pass the tests.
- [x] Test 1: Same swimlane | Same status
  - [x] Open __Kanban__ of __Project Example 1__ http://localhost:9001/project/project-1/kanban 
  - [x] Move user stories `#63` under `#24`
  - [x] Refresh the page and the swimlane `quos` status `new` should be `#24 #63 #47`
- [x] Test 2: Same swimlane | Different status
  - [x] Open __Kanban__ of __Project Example 1__ http://localhost:9001/project/project-1/kanban 
  - [x] Move user stories `#63` to the top of  `ready` status
  - [x] Refresh the page and the swimlane `quos` status `new` should be `#24 #47` and `ready` contains `#63` 
- [x] Test 3: Different swimlanes 
  - [x] Open __Kanban__ of __Project Example 1__ http://localhost:9001/project/project-1/kanban 
  - [x] Move user stories `#63` frpm `quas/ready` to `autem quas/in progress`
  - [x] Refresh the page and the swimlane `quos/ready` should be empty and `autem quas/in progress` should contains `#63`
- [x] Test 4: Play with it
  - [x] Do more tests on your own 
- [ ] Move [Task #762](https://tree.taiga.io/project/taiganext/task/762) to `ready-for-qa`

### CODE REVIEW

File: taiga/base/fields.py [hunk 1]
```diff
@@ -77,6 +77,75 @@ def from_native(self, data):
         return data
 
 
+class ListField(serializers.WritableField):
+    """
+    A field whose values are lists of items described by the given child. The child can
```
Comment (yamila-moreno) [inline taiga/base/fields.py:82]: eing?

File: taiga/projects/userstories/services.py [hunk 2]
```diff
@@ -112,6 +117,79 @@ def update_userstories_order_in_bulk(bulk_data: list, field: str,
...
+     - `bulk_userstories` should be a list of user stories IDs
+    """
+    # filter user stories from status and swimlane
+    user_stories = project.user_stories.filter(status=status)
+    if swimlane is not None:
+         user_stories = user_stories.filter(swimlane=swimlane)
+    else:
+         user_stories = user_stories.filter(swimlane__isnull=True)
+
+    # exclude moved user storiesi and
```
Comment (yamila-moreno) [inline taiga/projects/userstories/services.py:138]: typo stories

File: taiga/projects/userstories/services.py [hunk 3]
```diff
@@ -112,6 +117,79 @@ def update_userstories_order_in_bulk(bulk_data: list, field: str,
...
+    user_stories = user_stories.exclude(id__in=bulk_userstories)
+
+    # if after_userstory, exclude it and get only elements after it:
+    if after_userstory:
```
Comment (yamila-moreno) [inline taiga/projects/userstories/services.py:142]: si hay user_story y se filtra por __gte kanban_order no haría falta excluirla, no? el filtro ya la excluye ¿?

File: tests/integration/test_userstories_update_kanban_order.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+pytestmark = pytest.mark.django_db(transaction=True)
+
+
+##############################
+## Move to no swimlane
+##############################
+
+
+def test_api_update_orders_in_bulk_succeeds_moved_to_no_swimlane_and_to_the_begining(client):
+    #
```
Comment (yamila-moreno) [inline tests/integration/test_userstories_update_kanban_order.py:47]: aplausos por esta tabla tan bonica :+1:

File: tests/integration/test_userstories_update_kanban_order.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+    assert us2.status_id == status1.id
+    assert us3.kanban_order == 2
+    assert us3.swimlane_id == None
+    assert us3.status_id == status1.id
+    assert us1.kanban_order == 3
+    assert us1.swimlane_id == None
+    assert us1.status_id == status1.id
+
+
+def test_api_update_orders_in_bulk_succeeds_moved_to_no_swimlane_and_to_the_middle(client):
```
Comment (yamila-moreno) [inline tests/integration/test_userstories_update_kanban_order.py:105]: por confirmar, ¿se pretende que este test valide dos operaciones de reordenado? al pasar us3 a st1 no debería cambiar el orden relativo de us1 y us4

File: tests/integration/test_userstories_update_kanban_order.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+def test_api_update_orders_in_bulk_succeeds_moved_to_a_swimlane_and_to_the_end(client):
+    #
+    #      |  ST1  |  ST2    |               |       |  ST1  |  ST2
+    # -----|-------|-------  |               |  -----|-------|-------
+    #  SW1 |  us1  |         |   MOVE: us3   |       |  us1  |
+    #      |  us2  |         |   TO: sw1     |   SW1 |  us2  |
+    # -----|-------|-------  |   UNDER: st1  |       |  us3  |
+    #      |       |  us3    |   AFTER: us2  |  -----|-------|-------
+    #  SW2 |       |         |               |       |       |
+    #      |       |         |               |   SW1 |       |
```
Comment (yamila-moreno) [inline tests/integration/test_userstories_update_kanban_order.py:375]: typo SW1 > SW2

File: taiga/base/fields.py (hunk 1 above)
Comment (bameda) [reply taiga/base/fields.py:82]: What's the problem? This is something like this: https://www.django-rest-framework.org/api-guide/fields/#listfield

File: taiga/projects/userstories/services.py (hunk 3 above)
Comment (bameda) [reply taiga/projects/userstories/services.py:142]: Precisamente por el `e` de `gte` no se excluye, podría darse el caso de que `after_userstory` estuviese dentro de las que estas moviendo, pero como no quiero excluir aquellas con la misma ordenación (que se podría dar el caso de dos con el mismo orden) tengo que hacer la exclusión explicita para estar seguro

File: taiga/projects/userstories/services.py (hunk 2 above)
Comment (bameda) [reply taiga/projects/userstories/services.py]: Fixed

File: tests/integration/test_userstories_update_kanban_order.py (hunk 6 above)
Comment (bameda) [reply tests/integration/test_userstories_update_kanban_order.py]: Fixed

File: tests/integration/test_userstories_update_kanban_order.py (hunk 4 above)
Comment (bameda) [reply tests/integration/test_userstories_update_kanban_order.py:47]: :clap:

File: taiga/projects/userstories/services.py (hunk 3 above)
Comment (yamila-moreno) [reply taiga/projects/userstories/services.py:142]: tienes toda la razón, me colé, ¡gracias!

**Pull request merged**